controllers/movies.ts

```diff
@@ -0,0 +1,302 @@
+import { Movie } from "../model/Movie";
+import { shortenURLs } from "../shorten-url";
+import { MovieParams, MovieQuerableParams } from "../util/types";
+
+class Movies {
+  async getAllMovies(params: MovieQuerableParams) {
+    const { actor, imdbSort, yearSort } = params;
+    try {
+      let query = Movie.createQueryBuilder("movie");
+      if (actor) {
+        query = query.where("movie.Actors LIKE :actor", {
+          actor: `%${actor}%`,
+        });
+      }
+      if (imdbSort) {
+        const imdbSortDirection =
+          imdbSort.toUpperCase() === "ASC" ? "ASC" : "DESC";
+        query = query.orderBy("movie.imdbRating", imdbSortDirection);
+      }
+
+      if (yearSort) {
+        const yearSortDirection =
+          yearSort.toUpperCase() === "ASC" ? "ASC" : "DESC";
+        query = query.orderBy("movie.Year", yearSortDirection);
+      }
+      let movies;
+
+      if (imdbSort) {
+        movies = await query
+          .select(["movie.Title", "movie.imdbRating"])
+          .getMany();
+      } else {
+        movies = await query.getMany();
+      }
+      return movies;
+    } catch (err: any) {
+      throw new Error("Movies not found");
+    }
+  }
+
+  async getMovieByImdID(imdb: any) {
+    try {
+      const movie = await Movie.findOne({
+        where: {
+          imdbID: imdb,
+        },
+      });
+      return movie;
+    } catch (err: any) {
+      throw new Error("Movie not found");
+    }
+  }
+
+  async deleteMovie(id: any) {
+    try {
+      const movie = (await Movie.findOne({
+        where: {
+          Id: id,
+        },
+      })) as Movie;
+      await Movie.remove(movie);
+      return movie;
+    } catch (err: any) {
+      throw new Error("Movie not found");
+    }
+  }
+
+  async addMovie(params: MovieParams) {
+    const {
+      Title,
+      Year,
+      Rated,
+      Released,
+      Runtime,
+      Genre,
+      Director,
+      Writer,
+      Actors,
+      Plot,
+      Language,
+      Country,
+      Awards,
+      Poster,
+      Metascore,
+      imdbRating,
+      imdbVotes,
+      imdbID,
+      Type,
+      Response,
+      Images,
+      ComingSoon,
+    } = params;
+
+    try {
+      const movie = await Movie.create({
```

Comment:
If you are creating movie from with the same params you can just put the object params as attribute instead of destructuring. ex. await Movie.create(params)

Reply:
I got the error Argument of type 'MovieParams' is not assignable to parameter of type 'DeepPartial<Movie>[]      Type 'MovieParams' is missing the following properties from type 'DeepPartial<Movie>[]': length, pop, push, concat, and 29 more....   
 
but assigning the params to the obj movie worked, so i fixed it a bit different 
      Object.assign(movie, params);
      const newMovie = await movie.save();

Reply:
This error was most probably because of missing some required data of the Movie model which needs to be fixed. With your solution you are modifying the movie object and assigning params to it. Try to log the values in both way (with the fix and before) and you will find what is missing.